File: apps/api/app/core/answers.py
"""Нормализация и сравнение ответов.

Зеркало `packages/core/src/answers.ts`. Клиент показывает вердикт мгновенно,
сервер пересчитывает его заново при проверке теста и не доверяет клиенту.
Расхождение реализаций означало бы, что человек видит «верно», а в результатах
теста получает ошибку, поэтому обе проверяются одним набором кейсов —
`packages/core/src/answer-cases.json`.

Порядок шагов нормализации и таблицы допусков обязаны совпадать с TS до символа.
"""

from __future__ import annotations

import enum
import re
import unicodedata
from dataclasses import dataclass


class Strictness(enum.StrEnum):
    strict = "strict"
    moderate = "moderate"
    lenient = "lenient"


class AnswerVerdict(enum.StrEnum):
    correct = "correct"
    typo = "typo"
    incorrect = "incorrect"


# Апострофы удаляются, остальная пунктуация заменяется пробелом:
# «don't» → «dont», но «кошка,собака» → «кошка собака».
_APOSTROPHES = re.compile(r"['’ʼ`´]")
_PUNCTUATION = re.compile(r"[.,;:!?«»„“”\"()\[\]{}<>/\\|—–\-_…*+=~^&%$#@]")
_WHITESPACE = re.compile(r"\s+")

# Диакритика снимается только с латиницы. В кириллице «й» — это «и» с кратким,
# и слепая свёртка приравняла бы «мой» к «мои». «ё» приводится к «е» отдельно:
# это привычная замена, а не потеря буквы, и работает во всех режимах.
_LATIN_DIACRITICS = re.compile(r"([a-z])[̀-ͯ]+")

# Артикли отбрасываются только в начале ответа и только для не-русских языков.
_ARTICLES: dict[str, frozenset[str]] = {
    "en": frozenset({"a", "an", "the"}),
    "de": frozenset(
        {"der", "die", "das", "den", "dem", "des", "ein", "eine", "einen", "einem", "einer"}
    ),
    "fr": frozenset({"le", "la", "les", "un", "une", "des", "l'", "du", "de"}),
    "es": frozenset({"el", "la", "los", "las", "un", "una", "unos", "unas"}),
    "it": frozenset({"il", "lo", "la", "i", "gli", "le", "un", "uno", "una"}),
}

# Допуск на опечатку по длине ожидаемого ответа: (максимальная длина, расстояние).
# Таблицами, а не формулой — значения должны совпадать с TS до единицы.
_TYPO_THRESHOLDS: dict[Strictness, tuple[tuple[float, int], ...]] = {
    Strictness.strict: ((float("inf"), 0),),
    Strictness.moderate: ((3, 0), (7, 1), (float("inf"), 2)),
    Strictness.lenient: ((2, 0), (5, 1), (10, 2), (float("inf"), 3)),
}


@dataclass(frozen=True, slots=True)
class AnswerResult:
    verdict: AnswerVerdict
    matched: str | None
    distance: float


def normalize_answer(
    value: str, *, strictness: Strictness = Strictness.moderate, lang: str | None = None
) -> str:
    result = value.strip().lower().replace("ё", "е")

    if strictness is not Strictness.strict:
        decomposed = unicodedata.normalize("NFD", result)
        result = unicodedata.normalize("NFC", _LATIN_DIACRITICS.sub(r"\1", decomposed))
        result = _APOSTROPHES.sub("", result)
        result = _PUNCTUATION.sub(" ", result)

    result = _WHITESPACE.sub(" ", result).strip()

    if strictness is not Strictness.strict:
        result = _strip_article(result, lang)
    return result


def check_answer(
    typed: str,
    expected: str,
    *,
    strictness: Strictness = Strictness.moderate,
    alternatives: list[str] | None = None,
    lang: str | None = None,
) -> AnswerResult:
    candidates = [expected, *(alternatives or [])]
    normalized_typed = normalize_answer(typed, strictness=strictness, lang=lang)
    if not normalized_typed:
        return AnswerResult(AnswerVerdict.incorrect, None, float("inf"))

    best = AnswerResult(AnswerVerdict.incorrect, None, float("inf"))
    for candidate in candidates:
        normalized = normalize_answer(candidate, strictness=strictness, lang=lang)
        if not normalized:
            continue
        if normalized == normalized_typed:
            return AnswerResult(AnswerVerdict.correct, candidate, 0)
        distance = levenshtein(normalized_typed, normalized)
        if distance < best.distance:
            verdict = (
                AnswerVerdict.typo
                if distance <= typo_threshold(len(normalized), strictness)
                else AnswerVerdict.incorrect
            )
            best = AnswerResult(verdict, candidate, distance)
    return best


def typo_threshold(length: int, strictness: Strictness = Strictness.moderate) -> int:
    """Расстояние, при котором ответ ещё считается опечаткой, а не ошибкой."""
    for max_length, threshold in _TYPO_THRESHOLDS[strictness]:
        if length <= max_length:
            return threshold
    return 0


def levenshtein(left: str, right: str) -> int:
    if left == right:
        return 0
    if not left:
        return len(right)
    if not right:
        return len(left)

    previous = list(range(len(right) + 1))
    for i, left_char in enumerate(left, start=1):
        current = [i]
        for j, right_char in enumerate(right, start=1):
            substitution = previous[j - 1] + (0 if left_char == right_char else 1)
            current.append(min(current[j - 1] + 1, previous[j] + 1, substitution))
        previous = current
    return previous[len(right)]


def _strip_article(value: str, lang: str | None) -> str:
    articles = _ARTICLES.get((lang or "ru")[:2].lower())
    if not articles:
        return value
    head, separator, rest = value.partition(" ")
    # Односложный ответ не трогаем: «the» как ответ на «определённый артикль»
    # должен остаться собой.
    if not separator:
        return value
    return rest if head in articles else value

File: packages/core/src/answers.ts
/**
 * Нормализация и сравнение ответов для режимов «Письмо», «Тест» и «Аудирование».
 *
 * Логика продублирована на Python (`apps/api/app/core/answers.py`): клиент
 * показывает вердикт мгновенно, сервер проверяет ответы теста заново и не
 * доверяет клиенту. Расхождение двух реализаций означало бы, что человек видит
 * «верно», а в результатах теста получает ошибку, поэтому обе проверяются одним
 * набором кейсов — `answer-cases.json`.
 *
 * Отдельный статус «опечатка» существует, чтобы промах по клавише не ломал
 * расписание: ошибкой это не считается, но ответ нужно ввести заново.
 */

export type Strictness = 'strict' | 'moderate' | 'lenient';
export type AnswerVerdict = 'correct' | 'typo' | 'incorrect';

export const STRICTNESS_LEVELS = ['strict', 'moderate', 'lenient'] as const;

export const STRICTNESS_LABELS: Record<Strictness, string> = {
  strict: 'Строго',
  moderate: 'Умеренно',
  lenient: 'Мягко',
};

export interface AnswerOptions {
  strictness?: Strictness;
  /** Синонимы, которые тоже засчитываются (`alt_answers` карточки). */
  alternatives?: readonly string[];
  /** Язык ответа. Для не-русского отбрасываются артикли. */
  lang?: string;
}

export interface AnswerResult {
  verdict: AnswerVerdict;
  /** Вариант, с которым ответ совпал или к которому оказался ближе всего. */
  matched: string | null;
  /** Расстояние Левенштейна до ближайшего варианта после нормализации. */
  distance: number;
}

/**
 * Апострофы удаляются, остальная пунктуация заменяется пробелом:
 * «don't» → «dont», но «кошка,собака» → «кошка собака».
 */
const APOSTROPHES = /['’ʼ`´]/g;
const PUNCTUATION = /[.,;:!?«»„“”"()[\]{}<>/\\|—–\-_…*+=~^&%$#@]/g;
/**
 * Диакритика снимается только с латиницы. В кириллице «й» — это «и» с кратким,
 * и слепая свёртка приравняла бы «мой» к «мои». «ё» приводится к «е» отдельно:
 * это привычная замена, а не потеря буквы, и работает во всех режимах.
 */
const LATIN_DIACRITICS = /([a-z])[\u0300-\u036f]+/g;

/** Артикли отбрасываются только в начале ответа и только для не-русских языков. */
const ARTICLES: Record<string, readonly string[]> = {
  en: ['a', 'an', 'the'],
  de: ['der', 'die', 'das', 'den', 'dem', 'des', 'ein', 'eine', 'einen', 'einem', 'einer'],
  fr: ['le', 'la', 'les', 'un', 'une', 'des', "l'", 'du', 'de'],
  es: ['el', 'la', 'los', 'las', 'un', 'una', 'unos', 'unas'],
  it: ['il', 'lo', 'la', 'i', 'gli', 'le', 'un', 'uno', 'una'],
};

/**
 * Допуск на опечатку по длине ожидаемого ответа. Таблицами, а не формулой:
 * значения должны совпадать с Python до единицы, а формулу легко разойтись
 * в округлении.
 */
const TYPO_THRESHOLDS: Record<Strictness, readonly (readonly [number, number])[]> = {
  // [максимальная длина, допустимое расстояние]
  strict: [[Infinity, 0]],
  moderate: [
    [3, 0],
    [7, 1],
    [Infinity, 2],
  ],
  lenient: [
    [2, 0],
    [5, 1],
    [10, 2],
    [Infinity, 3],
  ],
};

export function normalizeAnswer(value: string, options: AnswerOptions = {}): string {
  const strictness = options.strictness ?? 'moderate';
  let result = value.trim().toLowerCase().replace(/ё/g, 'е');

  if (strictness !== 'strict') {
    result = result.normalize('NFD').replace(LATIN_DIACRITICS, '$1').normalize('NFC');
    result = result.replace(APOSTROPHES, '').replace(PUNCTUATION, ' ');
  }

  result = result.replace(/\s+/g, ' ').trim();

  if (strictness !== 'strict') {
    result = stripArticle(result, options.lang);
  }
  return result;
}

export function checkAnswer(
  typed: string,
  expected: string,
  options: AnswerOptions = {},
): AnswerResult {
  const strictness = options.strictness ?? 'moderate';
  const candidates = [expected, ...(options.alternatives ?? [])];
  const normalizedTyped = normalizeAnswer(typed, options);

  if (normalizedTyped.length === 0) {
    return { verdict: 'incorrect', matched: null, distance: Infinity };
  }

  let best: AnswerResult = { verdict: 'incorrect', matched: null, distance: Infinity };
  for (const candidate of candidates) {
    const normalized = normalizeAnswer(candidate, options);
    if (normalized.length === 0) continue;
    if (normalized === normalizedTyped) {
      return { verdict: 'correct', matched: candidate, distance: 0 };
    }
    const distance = levenshtein(normalizedTyped, normalized);
    if (distance < best.distance) {
      best = {
        verdict: distance <= typoThreshold(normalized.length, strictness) ? 'typo' : 'incorrect',
        matched: candidate,
        distance,
      };
    }
  }
  return best;
}

/** Допустимое расстояние, при котором ответ считается опечаткой, а не ошибкой. */
export function typoThreshold(length: number, strictness: Strictness = 'moderate'): number {
  for (const [maxLength, threshold] of TYPO_THRESHOLDS[strictness]) {
    if (length <= maxLength) return threshold;
  }
  return 0;
}

/** Расстояние Левенштейна. Итеративная версия на двух строках. */
export function levenshtein(left: string, right: string): number {
  if (left === right) return 0;
  if (left.length === 0) return right.length;
  if (right.length === 0) return left.length;

  let previous = Array.from({ length: right.length + 1 }, (_, index) => index);
  let current = new Array<number>(right.length + 1);

  for (let i = 1; i <= left.length; i += 1) {
    current[0] = i;
    for (let j = 1; j <= right.length; j += 1) {
      const substitution = previous[j - 1]! + (left[i - 1] === right[j - 1] ? 0 : 1);
      current[j] = Math.min(current[j - 1]! + 1, previous[j]! + 1, substitution);
    }
    [previous, current] = [current, previous];
  }
  return previous[right.length]!;
}

/** Процент посимвольного совпадения после той же нормализации, что и у проверки ответа. */
export function answerSimilarity(
  typed: string,
  expected: string,
  options: AnswerOptions = {},
): number {
  const left = normalizeAnswer(typed, options);
  const right = normalizeAnswer(expected, options);
  if (left.length === 0 || right.length === 0) return 0;
  const longest = Math.max(left.length, right.length);
  return Math.round((1 - levenshtein(left, right) / longest) * 100);
}

function stripArticle(value: string, lang: string | undefined): string {
  const articles = ARTICLES[(lang ?? 'ru').slice(0, 2).toLowerCase()];
  if (!articles) return value;
  const spaceAt = value.indexOf(' ');
  if (spaceAt === -1) return value;
  const head = value.slice(0, spaceAt);
  // Односложный ответ не трогаем: «the» как ответ на «определённый артикль»
  // должен остаться собой.
  return articles.includes(head) ? value.slice(spaceAt + 1) : value;
}
